dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/css/OpenAPIUIBundlesUpdater.java

```diff
@@ -0,0 +1,351 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.microprofile.openapi20.css;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleException;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.wsspi.kernel.service.utils.FileUtils;
+
+import io.openliberty.microprofile.openapi20.utils.CloudUtils;
+import io.openliberty.microprofile.openapi20.utils.Constants;
+import io.openliberty.microprofile.openapi20.utils.LoggingUtils;
+
+/**
+ * Update all OpenAPI-UI bundles
+ */
+public class OpenAPIUIBundlesUpdater {
+
+    private static final TraceComponent tc = Tr.register(OpenAPIUIBundlesUpdater.class);
+
+    /**
+     * List of OpenAPI-UI bundles
+     */
+    private static final Set<String> openAPIUIBundleNames = new HashSet<String>(Arrays.asList(new String[] {"com.ibm.ws.microprofile.openapi.ui"}));
+
+    private static boolean stopping = false;
+
+    synchronized static void updateResources(Map<String, Object> resourcesToUpdate, boolean isRestoreDefaults) throws IOException, BundleException {
```

Comment:
Is this method really rewriting the content of the UI bundle?

Why on earth is that necessary?

Reply:
This the implementation of the functionality that allows users to customise the OpenAPI user interface (Swagger UI), as described here:

https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_api_mpopenapi_custom.html

Reply:
Yeah, but we do that by rewriting the bundle content and restarting the bundles!? Just seems bizarre.

Reply:
It's a good question... and one that I do not know the answer to I am afraid.  This is value add functionality that goes beyond what is required by the MP OpenAPI spec.  I copied it across because I was told that we cannot remove functionality as part of adding support for MP OpenAPI 2.0.